### PR TITLE
Bug 1266116 - Create an Admin page.  A faster, full-page exclusion editor

### PR DESCRIPTION
**Pull request by camd**
This moves exclusion editing to a full-page Admin panel.  It also uses ReactJS to speed up the rendering of the long, long lists like ``job_types``.  Now the page can load in a few seconds instead of 5 or more minutes.

I apologize for the number of files modified.  Couldn't be helped.

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/treeherder/1821)

### CODE REVIEW

File: ui/js/controllers/admin/exclusions_detail.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
+'use strict';
```
Comment (camd) [inline ui/js/controllers/admin/exclusions_detail.js]: I couldn't decide whether it was better to leave the ``exclusions_detail.js`` and ``exclusions_list.js`` controllers separate or merge them.  I'm really fine going either way, if you have a preference.  Same for the ``profiles_list.js`` and ``profiles_detail.js`` controllers.

**Comment by wlach** [comment]
This looks great! I feel like ideally the sheriffing view would not be a seperate page, but we'd just swap the jobs list out if the user selected the admin mode. I'd be happy for that to be done as a followup though, since it would be rather involved (i.e. it would probably involve some deep changes to how we do client-side routing).

---

Review status: 0 of 37 files reviewed at latest revision, 12 unresolved discussions.

---

*[docs/ui/development.rst, line 9 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZA9B9Xc_gpK8O_v1z:-KRZA9B9Xc_gpK8O_v2-:b5bx8dl) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/docs/ui/development.rst#L9)):*
> ```reStructuredText
> 
> Modifying JSX files
> -------------------
> ```

This sounds like a major gotcha. Is there no way we could hook up web-server.js to automatically do this before us? It's just a node app so it feels like it should be possible.

---

*[docs/ui/development.rst, line 18 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZ9r3HsKa1CtmUbuUj:-KRZ9r3IyiTyW5xIOxx0:bgv9zqn) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/docs/ui/development.rst#L18)):*
> ```reStructuredText
> This will start a "watcher" that will continuously recompile any ``JSX`` file
> in the ``ui/js/react_components`` folder.  They are compiled to a single
> file called ``ui/js/react-compiled.js``.  Every chang you make to any ``JSX``
> ```

Every chang -> Every change

---

*[ui/js/controllers/admin.js, line 12 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZByi_Rk-OwT90P0YW:-KRZByi_Rk-OwT90P0YX:b-dir6ru) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/controllers/admin.js#L12)):*
> ```JavaScript
>         $q,ThExclusionProfileModel) {
> 
>         $scope.REFRESH_MSG = "Refresh Treeherder windows to see changes reflected.";
> ```

If you can avoid it,  I think it's best not to define view messages inside controllers.

---

*[ui/js/controllers/admin.js, line 59 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZBegvR1YdjbLLhKjA:-KRZBegw8H2_JQUA3de1:b-19ljab) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/controllers/admin.js#L59)):*
> ```JavaScript
>                 $scope.master_platforms = [];
>                 var platformPromise = ThBuildPlatformModel.get_list()
>                     .then(function(data) {
> ```

I'd prefer if the callback variable was buildPlatforms, not data. More descriptive.

---

*[ui/js/controllers/admin.js, line 62 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZAaXaaOrQ0CoooTml:-KRZAaXbKXOntm3QJTYr:b-uc7jtc) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/controllers/admin.js#L62)):*
> ```JavaScript
>                         for (var i = 0; i < data.length; i++) {
>                             $scope.master_platforms.push(
>                                 getJobComboField(data[i].platform, data[i].architecture)
> ```

We could simplify this a bunch by using map and ES6 template literals
```js
$scope.master_platforms = _.uniq(data.map(function(buildPlatform) { return`${buildPlatform[i].platform} (${buildPlatform[i].architecture})`); }).sort());
```

---

*[ui/js/controllers/admin.js, line 77 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZBlfYkvycDWq0Z_DK:-KRZBlfYkvycDWq0Z_DL:b9zvr7x) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/controllers/admin.js#L77)):*
> ```JavaScript
>                                 getJobComboField(data[i].name, data[i].symbol)
>                             );
>                         }
> ```

You could simplify things using a similar strategy to what I outlined above.

---

*[ui/js/controllers/admin/exclusions_detail.js, line 61 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZC_9b1MxIqmLBWwJW:-KRZC_9c4XyJYs4dW57O:bapjeve) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/controllers/admin/exclusions_detail.js#L61)):*
> ```JavaScript
>                 }, null);
>             }
>             thNotify.send($scope.REFRESH_MSG);
> ```

I'm not crazy about using thNotify to communicate the state here. Is there some other more intuitive way we can signify to the user that something has been save?

---

*[ui/js/controllers/admin/profiles_detail.js, line 53 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZCsKSo97gJEBee4Pm:-KRZCsKSo97gJEBee4Pn:b-e2l293) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/controllers/admin/profiles_detail.js#L53)):*
> ```JavaScript
>                     choices.push(key);
>                 }
>             });
> ```

Again, a simple map function would be better here, I think.

---

*[ui/js/directives/treeherder/main.js, line 186 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZDJBoeyrxBxYWBHX2:-KRZDJBoeyrxBxYWBHX3:b-xr191v) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/directives/treeherder/main.js#L186)):*
> ```JavaScript
>                 rightList: "="
>             },
>             link: function(scope, elem){
> ```

It might be nice to turn this into an angular component while you're here (maybe as a seperate PR), since they are so much nicer and easier to understand.

---

*[ui/partials/admin/exclusions_list.html, line 10 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZDn3uFHU8c8_g9gSO:-KRZDn3v6xg0VvfdESHR:b-fitppl) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/partials/admin/exclusions_list.html#L10)):*
> ```HTML
>             class="btn btn-sm btn-primary action-buttons">Add exclusion</button>
> 
>     <p ng-if="exclusions.length == 0">No exclusion available</p>
> ```

exclusions.length== 0 -> !exclusions.length

---

*[ui/partials/admin/exclusions_list.html, line 11 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZE0q2VinwumGkOLv0:-KRZE0q2VinwumGkOLv1:b-9cj2e0) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/partials/admin/exclusions_list.html#L11)):*
> ```HTML
> 
>     <p ng-if="exclusions.length == 0">No exclusion available</p>
>     <table ng-if="exclusions.length > 0"
> ```

exclusions.length > 0 -> exclusions.length

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-:-KRZEGR9Ovu2M5Qo3VBZ:b-de271r)*

**Comment by edmorley** [comment]
Awesome work!

Some higher-level thoughts/questions:
* Should we commit the compiled JSX content, or build dynamically? I'm leaning towards the latter, given that's what we do for the rest of the UI build.
* Should we use JSX or just plain JS (you don't have to use JSX when using React). The latter would simplify things in terms of build/dev pipeline for now. If/when we start to use React more (and by which time we're on Heroku with newer nodejs etc), we could always add JSX support at that point. Although since we already use Babel, perhaps there is little extra complexity?

File: Gruntfile.js [hunk 2]
```diff
@@ -65,10 +65,17 @@ module.exports = function(grunt) {
...
+                nonull: true,
+                options:{
+                    dest:'dist'
+                }
+            },
         },
 
         usemin:{ html:['dist/index.html', 'dist/userguide.html', 'dist/logviewer.html',
-                       'dist/perf.html', 'dist/failureviewer.html'] },
+                       'dist/perf.html', 'dist/failureviewer.html', 'dist/admin.html'] },
```
Comment (edmorley) [inline Gruntfile.js]: Could you try using something like `['dist/*.html']`? (In a separate, earlier commit)

File: Gruntfile.js [hunk 3]
```diff
@@ -159,6 +178,14 @@ module.exports = function(grunt) {
                 nonull: true,
                 flatten: true
                 },
+            // Copy vendor files that don't work with grouped minification
+            vendor:{
+                files: [
+                    { src: 'ui/vendor/ngReact/ngReact.min.js', dest: 'dist/vendor/ngReact/ngReact.min.js', nonull: true },
+                    { src: 'ui/vendor/react/react.min.js', dest: 'dist/vendor/react/react.min.js', nonull: true },
+                    { src: 'ui/vendor/react/react-dom.min.js', dest: 'dist/vendor/react/react-dom.min.js', nonull: true },
```
Comment (edmorley) [inline Gruntfile.js]: These can be copied with a wildcard I think?

**Comment by camd** [comment]
I considered the idea of making this full-page in place of the jobs view, but then thought better of it.  It feels like this will be shoe-horning the admin panel into a spot where it doesn't really belong.  If you replace the jobs view with this panel, then you'll have lots of menu items up top that don't apply (Infra, Repos, Filters, Unclassified count, etc.).  Those don't have anything to do with the process of updating the job exclusions.  The normal workflow for sheriffs is to have multiple tabs open to different repos, and supplanting one of them with the Admin page doesn't seem to buy you anything.  It feels, to me anyway, that it's cleaner to just open a new tab and not interrupt the context of the job view tab.  

I think a common (most common?) scenario would be a Sheriff in a TH tab and sees some jobs that need hiding.  If the admin panel replaces the current job view, and the Sheriff needs to go back to that to reference exactly which jobs to hide, they'd need to open another tab anyway to view where they just were.  So it seems that opening a new tab with just the applicable controls is the cleanest way to handle this.

But, of course, this is just my opinion.  :)  I welcome feedback on that.

---

Review status: 0 of 37 files reviewed at latest revision, 14 unresolved discussions.

---

*[Gruntfile.js, line 78 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KQS2dnjlhF03FH43tDf-r6-78:-KR_MG3_elNC7HzriEmt:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/Gruntfile.js#L78)):*
<details><summary><i>Previously, edmorley (Ed Morley) wrote…</i></summary>
> Could you try using something like `['dist/*.html']`? (In a separate, earlier commit)
</details>
Done.

---

*[Gruntfile.js, line 186 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KQS2dnjlhF03FH43tDf-r6-186:-KR_MDmth9PFy5ICd1lw:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/Gruntfile.js#L186)):*
<details><summary><i>Previously, edmorley (Ed Morley) wrote…</i></summary>
> These can be copied with a wildcard I think?
</details>
Done.

---

*[docs/ui/development.rst, line 9 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZA9B9Xc_gpK8O_v1z:-KRdEfOtjdR6_Be2YTxe:bwo3obb) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/docs/ui/development.rst#L9)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> This sounds like a major gotcha. Is there no way we could hook up web-server.js to automatically do this before us? It's just a node app so it feels like it should be possible.
</details>
I think there is a way to do this, yeah.  I'll look into it.  Maybe in a follow-up PR?  I think it's still good to have this ``build-react`` directive for when you're using ``runserver`` or ``gunicorn`` instead.  And I think I'll also have it called in the ``grunt build`` as Ed suggested.

---

*[docs/ui/development.rst, line 18 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZ9r3HsKa1CtmUbuUj:-KR_MU5hsjZinNceO6Rm:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/docs/ui/development.rst#L18)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> Every chang -> Every change
</details>
Done.

---

*[ui/js/controllers/admin.js, line 59 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZBegvR1YdjbLLhKjA:-KR_cepqYXtS4zEsDO9g:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/controllers/admin.js#L59)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> I'd prefer if the callback variable was buildPlatforms, not data. More descriptive.
</details>
Done.

---

*[ui/js/controllers/admin.js, line 62 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZAaXaaOrQ0CoooTml:-KR_cXBOxUhj4s9vf9U9:bafza4) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/controllers/admin.js#L62)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> We could simplify this a bunch by using map and ES6 template literals
> ```js
> $scope.master_platforms = _.uniq(data.map(function(buildPlatform) { return`${buildPlatform[i].platform} (${buildPlatform[i].architecture})`); }).sort());
> ```
</details>
Good call, yeah.  I had just copied this from the old Sheriff panel controller.  But this is nicer.

---

*[ui/js/controllers/admin.js, line 77 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZBlfYkvycDWq0Z_DK:-KR_cgYjQWAbvzNW7lPF:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/controllers/admin.js#L77)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> You could simplify things using a similar strategy to what I outlined above.
</details>
Done.

---

*[ui/js/controllers/admin/exclusions_detail.js, line 61 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZC_9b1MxIqmLBWwJW:-KR_NgQi0-4QgwgdWNjk:b-oh40p2) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/controllers/admin/exclusions_detail.js#L61)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> I'm not crazy about using thNotify to communicate the state here. Is there some other more intuitive way we can signify to the user that something has been save?
</details>
This is consistent with  the current UI pattern for the existing Sheriff panel as well as classifying jobs.  Did you have something better in mind?  Maybe if we come up with something better, we could address it as a whole across our UI.

---

*[ui/js/controllers/admin/profiles_detail.js, line 53 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZCsKSo97gJEBee4Pm:-KRdKW67qdgOicXa90S3:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/controllers/admin/profiles_detail.js#L53)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> Again, a simple map function would be better here, I think.
</details>
Done.

---

*[ui/js/directives/treeherder/main.js, line 186 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZDJBoeyrxBxYWBHX2:-KRdKY7DmdTPida3gmOb:b-rj4lvp) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/directives/treeherder/main.js#L186)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> It might be nice to turn this into an angular component while you're here (maybe as a seperate PR), since they are so much nicer and easier to understand.
</details>
Sure, I wrote a follow-up bug for it: https://bugzilla.mozilla.org/show_bug.cgi?id=1302757

---

*[ui/partials/admin/exclusions_list.html, line 10 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZDn3uFHU8c8_g9gSO:-KR_OuPDR8rbG_8BfPDF:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/partials/admin/exclusions_list.html#L10)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> exclusions.length== 0 -> !exclusions.length
</details>
Done.

---

*[ui/partials/admin/exclusions_list.html, line 11 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZE0q2VinwumGkOLv0:-KR_OuvNiC8_GJfwTVy7:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/partials/admin/exclusions_list.html#L11)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> exclusions.length > 0 -> exclusions.length
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-:-KRdCbcNmQVqXjhCUDoh:b-9a71pm)*

**Comment by camd** [comment]
After some digging and head-against-keyboard-banging, I've decided I'm going to skip using JSX for now.   It's proving quite hard to do this with grunt as well as compiling the JSX in the browser.  In the interest of getting this feature out quickly, I'm going to save off what I've done in that regard to another branch and make this one a simpler implementation.  We can decide later if we really DO want to use JSX or not.  Maybe we won't.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-:-KRdM2EWzJInuV8akSKY:b-wrafn7)*

**Comment by camd** [comment]
Review status: 0 of 33 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[ui/js/controllers/admin.js, line 12 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZByi_Rk-OwT90P0YW:-KRjhWtsPMlRiBqm8IYi:b-wuvph5) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/ui/js/controllers/admin.js#L12)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary>
> If you can avoid it,  I think it's best not to define view messages inside controllers.
</details>
I added this to the ``values.js`` as a string instead.

---

*[docs/ui/development.rst, line 9 at r6](https://reviewable.io:443/reviews/mozilla/treeherder/1821#-KRZA9B9Xc_gpK8O_v1z:-KRjhgnwtDo1GBHmYrij:biswyrj) ([raw file](https://github.com/mozilla/treeherder/blob/e53d080b71756040b87b3add752bd566bc3ab9e2/docs/ui/development.rst#L9)):*
<details><summary><i>Previously, camd wrote…</i></summary>
> I think there is a way to do this, yeah.  I'll look into it.  Maybe in a follow-up PR?  I think it's still good to have this ``build-react`` directive for when you're using ``runserver`` or ``gunicorn`` instead.  And I think I'll also have it called in the ``grunt build`` as Ed suggested.
</details>
As I mentioned in a general comment, I decided to remove the use of JSX for this PR.  We may want to use it later, but can cross that bridge when we come to it, since this instance is such a simple case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1821)*

**Pull request merged**